### PR TITLE
chore: enable cucumber, rewrite existing tests

### PR DESCRIPTION
**Pull request by Mohammer5**
This PR covers several topics

## Cucumber

I've added the `cypress-cucumber-preprocessor` npm module.
This allows us to write `.feature` files in the `cypress/integration` folder.

I've set `"nonGlobalStepDefinitions": true` in the `package.json` to enable us to place the step definitions in folders with the same name as the `.feature` file

## Rewrite existing tests

I've rewritten the previously existing tests for the AlertBar.
There are now two feature files in `cypress/integration`:
* `AlertBar/Autohiding.feature`
Step definitions in ``cypress/integration/AlertBar/Autohiding/`
* `AlertBar/ClearOnAction.feature`
Step definitions in ``cypress/integration/AlertBar/ClearOnAction/`

## Add scripts

I've added 5 scripts to the package.json of which 2 are of importance to us:
* `cypress:open`
* `cypress:run`

These will both run the storybook and then cyrpress.
What to check here:
(When running `yarn cypress:run`, make sure your BROWSER env var contains the path to either chrome or chromium: `BROWSER=$(which chrome) yarn cypress:run`)

1. Correct behavior
   * `yarn cypress:open` should simply open the cypress interface
   * `yarn cypress:run` should run the tests inside the terminal
   I've tested it on a different tty to ensure it's working properly because in ttys with a window manager, it still opens a real browser
   * When either of the two scripts are terminated or  `cypress:open` process is terminated manually:
   There should be no running process of the storybook or cypress
   * `cypress:run` should exit with code 0 if all tests pass
1. Error management
   * the storybook port is already in use
   The script should exit with code 1
   * the tests did not pass successfully
   The `cypress:run` script should exit with code 1

### CODE REVIEW

**Comment by Mohammer5** [comment]
Lacks docs and proper PR description.
Will add them for better understanding

**Comment by Mohammer5** [comment]
> Some file have some complex looking logic in them, i.e. cypress/assets/unfetch.umd.js and possibly this is something we'd need to copy into every repo that uses Gherkin. Perhaps we can use some form of package to take care of this? Or perhaps keep all these scripts in a central location (@dhis2/cli or sth)?

Unfetch is only temporarily needed until cypress supports hooking into fetch.
@amcgee and I will create a wrapper library that will take care of installing and setting up all of this in app platform apps (I've added a PR for setup script in `@dhis2/cli` but we both prefer an isolated library that will also mange the cypress and cucumber version)

**Comment by amcgee** [comment]
@Mohammer5 why do we need `unfetch` here?  I don't think `ui-core` components should be doing fetches?

In cypress tests for *applications* we should be able to simply `delete window.fetch` (I did this in [maps app](https://github.com/dhis2/maps-app/blob/master/cypress/support/network-fixtures.js#L13-L17))and then rely on the app's polyfill to stub it out, I think....  am I missing something here?

**Comment by Mohammer5** [comment]
> @Mohammer5 why do we need unfetch here? I don't think ui-core components should be doing fetches?
> 
> In cypress tests for applications we should be able to simply delete window.fetch (I did this in maps app)and then rely on the app's polyfill to stub it out, I think.... am I missing something here?

You're right, we don't need it in `ui-core` at all.. 
`ui-widgets` does need it though as there's no polyfill in the storybook (that's where I copied it from).

I'll remove it from this PR

**Comment by amcgee** [comment]
Nice!

I think it also shouldn't be necessary in `ui-widgets` though right?  Because the `ui-widgets` storybook uses a [CustomDataProvider](https://github.com/dhis2/ui-widgets/blob/master/stories/HeaderBar.stories.js#L114) which intercepts all data requests, so they shouldn't turn into fetches either.

File: .eslintrc.js [hunk 1]
```diff
@@ -3,4 +3,9 @@ const { config } = require('@dhis2/cli-style')
 module.exports = {
     parser: 'babel-eslint',
     extends: ['eslint:recommended', 'plugin:react/recommended', config.eslint],
+    env: { es6: true },
+    globals: {
+        Cypress: 'readonly',
```
Comment (amcgee) [inline .eslintrc.js]: These should only apply to the `cypress` folder, if possible.  Can we put a subsidiary `.eslintrc` in `cypress/`?  That might not be supported by cli-style...  @varl?

File: cypress.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
-  "screenshots": {
-    "screenshotOnRunFailure": false
-  },
-  "projectId": "65oh1u"
+    "testFiles": "**/*.feature",
+    "screenshots": {
+        "screenshotOnRunFailure": false
```
Comment (amcgee) [inline cypress.json]: Why disable this?

**Comment by Mohammer5** [comment]
Nice!

I think it also shouldn't be necessary in ui-widgets though right? Because the ui-widgets storybook uses a CustomDataProvider which intercepts all data requests, so they shouldn't turn into fetches either.

I'm using that to test different scenarios, like checking what happens when there are messages and when there are not messages. 
Alternatively we could add another scenario to the storybook instead if that's what should be preferred. I kinda liked the way how I was able to override certain responses though:

[Setting the defaults](https://github.com/dhis2/ui-widgets/blob/master/cypress/integration/HeaderBar/common/index.js)
[Overwriting the defaults](https://github.com/dhis2/ui-widgets/blob/master/cypress/integration/HeaderBar/Messages/index.js)

It works quite well with gherkin as well, as there'd be nothing to be done by the `Given` step definitions if it was done by the story already

File: cypress/fixtures/example.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (amcgee) [inline cypress/fixtures/example.json]: Probably should remove this example fixture if we're not using it

File: cypress.json (hunk 2 above)
Comment (Mohammer5) [reply cypress.json]: I think this should be explicitly turned on for CI. During development this won't be needed as we're probably debugging in the browser anyways

File: .eslintrc.js (hunk 1 above)
Comment (Mohammer5) [reply .eslintrc.js]: Oh yeah! That'd be awesome if supported

**Comment by Mohammer5** [comment]
>  what is the reasoning behind everything in scripts/cypress

I've tried to use concurrently and wait-on. Unfortunately wait-on doesn't support waiting for a 200 response, so it'll continue earlier than when storybook is ready. (See ["Goals"](https://github.com/jeffbski/wait-on#goals))

The scripts also make sure that the script fails if the port used by both cypress and storybook is identical, it exits with 1 if the port is already in use. Storybook just starts with an incremented port if the specified one is already in use and there's no way to know about that except for analyzing the output.

> I'm also not sure what cypress:ci is for, why isn't that just cypress:run?

`cypress:ci` is used internally by the script that's called from `cypress:run`:
* `cypress:run` starts the `ci` process
* `cypress:ci` is the process that expects the storybook to be available

I guess the naming is a bit odd here, I'd appreciate alternative names that are more descriptive!

File: cypress/fixtures/example.json (hunk 3 above)
Comment (Mohammer5) [reply cypress/fixtures/example.json]: Removed

**Comment by amcgee** [comment]
> I've tried to use concurrently and wait-on. Unfortunately wait-on doesn't support waiting for a 200 response, so it'll continue earlier than when storybook is ready. (See "Goals")

Do we need to wait for a `200` response or just a `GET` response?  With webpack-dev-server this was solved by using `wait-on http-get://....`, but maybe storybook's server responds with non-200 to GET when it isn't ready?  That would surprise me, but possible.

**Comment by Mohammer5** [comment]
> Do we need to wait for a 200 response or just a GET response? With webpack-dev-server this was solved by using wait-on http-get://...., but maybe storybook's server responds with non-200 to GET when it isn't ready? That would surprise me, but possible.

I haven't tried `http-get` protocol. Will do and see if that works

**Comment by amcgee** [comment]
> The scripts also make sure that the script fails if the port used by both cypress and storybook is identical, it exits with 1 if the port is already in use. Storybook just starts with an incremented port if the specified one is already in use and there's no way to know about that except for analyzing the output.

Just throwing out possibly-simpler solutions here, since I'm worried that adding a bunch of custom development scripts to a single repo (though I know we'll extract them into a wrapper lib eventually):

Could we use [detect-port](https://www.npmjs.com/package/detect-port)'s CLI to either fail if the port isn't available or pass a random available port to Storybook and on to Cypress?

**Comment by amcgee** [comment]
> I haven't tried http-get protocol. Will do and see if that works

Found that after a lot of frustration and debugging with this same issue in maps, it's handy!

File: cypress.json (hunk 2 above)
Comment (varl) [reply cypress.json]: Why remove the `projectId` for the Cypress Dashboard?

**Comment by Mohammer5** [comment]
Thanks to @amcgee I was able to get rid of the scripts folder and do everything with concurrently and wait-on :+1:.
Please review again!

File: .eslintrc.js (hunk 1 above)
Comment (varl) [reply .eslintrc.js]: @Mohammer5 Any success with this?

Comment (Mohammer5) [reply .eslintrc.js]: Yes! It's working!

File: cypress.json (hunk 2 above)
Comment (Mohammer5) [reply cypress.json]: It's just at the bottom of the file

**Comment by Mohammer5** [comment]
@varl @amcgee I think this is ready for another review

**Comment by varl** [comment]
LGTM.

As I mentioned in the other Cypress/Cucumber PR: I don't have experience setting it up so basing my review on how it looks and works. If there are any improvements to be made they can be done incrementally.

**Comment by varl** [comment]
Would be great if you could take another look @amcgee.

File: cypress/fixtures/example.json (hunk 3 above)
Comment (amcgee) [reply cypress/fixtures/example.json]: It looks like this file is still here?

File: cypress.json [hunk 4]
```diff
@@ -1,6 +1,5 @@
 {
-  "screenshots": {
-    "screenshotOnRunFailure": false
-  },
-  "projectId": "65oh1u"
+    "testFiles": "**/*.feature",
+    "video": false,
```
Comment (amcgee) [inline cypress.json]: Why are we setting `video: false` here?  Seems like we want videos when doing recorded CI runs...

File: cypress/fixtures/example.json (hunk 3 above)
Comment (Mohammer5) [reply cypress/fixtures/example.json]: No idea what happened there.. It's gone now

File: cypress.json (hunk 4 above)
Comment (Mohammer5) [reply cypress.json]: Same reason as for the screenshots: I think this should be explicitly turned on for CI. During development this won't be needed as we're probably debugging in the browser anyways.

These files will be git-ignored, so on the one hand devs won't necessarily notice these files have been created and on the other hand is the amount of disk space used by videos quite a bit.

Comment (amcgee) [reply cypress.json]: Ah you're right, totally fair!  I'd used the `--video false` flag for that, but putting it in the config file is definitely better if we can override it explicitly when running in CI.  Nice.

Comment (Mohammer5) [reply cypress.json]: Ok.. seems like right now that's not an option: kimmobrunfeldt/concurrently#33
So I'm inclined to remove these options from cypress.json until this 3.5 year old issue has been solved..... Do you agree?

Comment (amcgee) [reply cypress.json]: https://docs.cypress.io/guides/guides/environment-variables.html#Setting

Looks like an env bar would work too?

Comment (amcgee) [reply cypress.json]: (Untested)

Comment (Mohammer5) [reply cypress.json]: It's a bit more complicated. We'd have to set the config with code, depending on an env var, and I wasn't successful in doing so (I didn't try extensively though).

I think for now it's ok to remove it from the config and do this on a new/another branch/PR

Comment (amcgee) [reply cypress.json]: This works (with `"video": false` in `cypress.json`)

```
CYPRESS_VIDEO=true cypress run
```

I think you're right, it's better not to create screenshots or videos on a dev machine so let's default those to **false** in `cypress.json` (as you had done) and override them with the env vars in `.travis.yml`

Comment (amcgee) [reply cypress.json]: (no configuration with code required)

Comment (Mohammer5) [reply cypress.json]: I've added this functionality for taking screenshots manually as that didn't work.
Works for videos out of the box.

**Pull request merged**